### ISSUE TITLE
[data] `Dataset.split()` does not use `ray_remote_args`

### ISSUE THREAD

**Issue by krfricke** [ray-project/ray#39057]
### What happened + What you expected to happen

When a Dataset is configured to use specific `ray_remote_args`, these are not propagated to the `split()` operation.

In the source code, the split operation is retrieved with default resources (1 CPU):

```
    split_single_block = cached_remote_fn(_split_single_block)
```

https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/split.py#L162

This leads to problems e.g. in gradient-based trainers. XGBoostTrainer and LightGBMTrainer use "empty" head trainers, i.e. the Trainer class is instantiated without any resources. It is that trainer class that executes the dataset eventually. This will schedule an extra task, which will take up one more cluster CPU. In clusters that use the maximum number of available CPUs for XGBoost workers, this can lead to a deadlock.

### Versions / Dependencies

master

### Reproduction script

```
from typing import Tuple

import ray
from ray import tune
from ray.data import Dataset
from ray.train.xgboost import XGBoostTrainer
from ray.train import Result, ScalingConfig


def prepare_data() -> Tuple[Dataset, Dataset]:
    dataset = ray.data.read_csv("s3://anonymous@air-example-data/breast_cancer.csv", ray_remote_args={"num_cpus": 0})
    train_dataset, valid_dataset = dataset.train_test_split(test_size=0.3)
    return train_dataset, valid_dataset


def train_xgboost(num_workers: int, use_gpu: bool = False) -> Result:
    train_dataset, valid_dataset = prepare_data()

    # XGBoost specific params
    params = {
        "tree_method": "approx",
        "objective": "binary:logistic",
        "eval_metric": ["logloss", "error"],
    }

    trainer = XGBoostTrainer(
        scaling_config=ScalingConfig(num_workers=num_workers, use_gpu=use_gpu, trainer_resources={"CPU": 1}),
        label_column="target",
        params=params,
        datasets={"train": train_dataset, "valid": valid_dataset},
        num_boost_round=100,
    )
    tuner = tune.Tuner(trainer)
    result = tuner.fit()[0]
    print(result.metrics)

    return result



ray.init(num_cpus=3)
result = train_xgboost(num_workers=3, use_gpu=False)
```

Workaround:

```
class MyXGBoostTrainer(XGBoostTrainer):
    def __init__(self, *args, **kwargs):
        from ray.data._internal.split import _split_single_block
        from ray.data._internal.remote_fn import cached_remote_fn

        cached_remote_fn(_split_single_block, num_cpus=0)

        super().__init__(*args, **kwargs)
```

and then use `MyXGBoostTrainer` instead.

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by raulchen** [comment]
`ray.data.read_csv("s3://anonymous@air-example-data/breast_cancer.csv", ray_remote_args={"num_cpus": 0})` the `ray_remote_args` parameter is supposed to only set the read tasks. So the behavior is expected.
But we indeed don't have an API for setting the resources of `split` tasks.
